src/signing/petition_types/petition_entity.rs

```diff
@@ -0,0 +1,557 @@
+            intent_hash,
+            entity,
+            PetitionFactors::new_threshold(matrix.threshold_factors, matrix.threshold as i8),
+            PetitionFactors::new_override(matrix.override_factors),
+        )
+    }
+
+    pub fn new_unsecurified(
+        intent_hash: IntentHash,
+        entity: AddressOfAccountOrPersona,
+        instance: HierarchicalDeterministicFactorInstance,
+    ) -> Self {
+        Self::new(
+            intent_hash,
+            entity,
+            PetitionFactors::new_unsecurified(instance),
+            None,
+        )
+    }
+
+    /// Returns `true` signatures requirement has been fulfilled, either by
+    /// override factors or by threshold factors
+    pub fn has_signatures_requirement_been_fulfilled(&self) -> bool {
+        self.status() == PetitionFactorsStatus::Finished(PetitionFactorsStatusFinished::Success)
+    }
+
+    fn union_of<F, T>(&self, map: F) -> IndexSet<T>
+    where
+        T: Eq + std::hash::Hash + Clone,
+        F: Fn(&PetitionFactors) -> IndexSet<T>,
+    {
+        self.both(
+            |l| map(l),
+            |t, o| {
+                t.unwrap_or_default()
+                    .union(&o.unwrap_or_default())
+                    .cloned()
+                    .collect::<IndexSet<T>>()
+            },
+        )
+    }
+
+    pub fn all_factor_instances(&self) -> IndexSet<OwnedFactorInstance> {
+        self.union_of(|l| l.factor_instances())
+            .into_iter()
+            .map(|f| OwnedFactorInstance::owned_factor_instance(self.entity.clone(), f.clone()))
+            .collect::<IndexSet<_>>()
+    }
+
+    pub fn all_skipped_factor_instance(&self) -> IndexSet<HierarchicalDeterministicFactorInstance> {
+        self.union_of(|f| f.all_skipped())
+    }
+
+    pub fn all_skipped_factor_sources(&self) -> IndexSet<FactorSourceIDFromHash> {
+        self.all_skipped_factor_instance()
+            .into_iter()
+            .map(|f| f.factor_source_id)
+            .collect::<IndexSet<_>>()
+    }
+
+    pub fn all_signatures(&self) -> IndexSet<HDSignature> {
+        self.union_of(|f| f.all_signatures())
+    }
+
+    fn with_list<F, T>(list: &Option<RefCell<PetitionFactors>>, map: F) -> Option<T>
+    where
+        F: Fn(&PetitionFactors) -> T,
+    {
+        list.as_ref().map(|refcell| map(&refcell.borrow()))
+    }
+
+    fn on_list<F, R>(&self, kind: FactorListKind, r#do: &F) -> Option<R>
+    where
+        F: Fn(&PetitionFactors) -> R,
+    {
+        match kind {
+            FactorListKind::Threshold => Self::with_list(&self.threshold_factors, r#do),
+            FactorListKind::Override => Self::with_list(&self.override_factors, r#do),
+        }
+    }
+
+    fn both<F, C, T, R>(&self, r#do: F, combine: C) -> R
+    where
+        F: Fn(&PetitionFactors) -> T,
+        C: Fn(Option<T>, Option<T>) -> R,
+    {
+        let t = self.on_list(FactorListKind::Threshold, &r#do);
+        let o = self.on_list(FactorListKind::Override, &r#do);
```

Comment:
so I'm building up utility of using logic on both lists. But since both lists are `Option<RefCell<PetitionForFactors>>` they are quite hard to use directly. 

`add_signature` calls `add_signature_if_relevant` on both, and then validates that we only added the signature to one list and only once.

I think perhaps best if you do a PR rewriting this if you have ideas on simplification? :)